pkg/service/v0/graph.go

```diff
@@ -0,0 +1,202 @@
+
+		if userID := chi.URLParam(r, "userID"); userID != "" {
+			user, err = g.ldapGetUser(userID)
+		} else {
+			// TODO: we should not give this error out to users
+			// http.Error(w, err.Error(), http.StatusInternalServerError)
+			render.Status(r, http.StatusNotFound)
+			return
+		}
+		if err != nil {
+			render.Status(r, http.StatusNotFound)
+			return
+		}
+
+		ctx := context.WithValue(r.Context(), userIDKey, user)
+		next.ServeHTTP(w, r.WithContext(ctx))
+	})
+}
+
+// GetMe implements the Service interface.
+func (g Graph) GetMe(w http.ResponseWriter, r *http.Request) {
+	me := createUserModel(
+		"Alice",
+		"1234-5678-9000-000",
+	)
+
+	resp, err := json.Marshal(me)
+
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	render.Status(r, http.StatusOK)
+	render.JSON(w, r, resp)
+}
+
+// GetUsers implements the Service interface.
+func (g Graph) GetUsers(w http.ResponseWriter, r *http.Request) {
+	con, err := g.initLdap()
+	if err != nil {
+		// TODO: we should not give this error out to users
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	result, err := g.ldapSearch(con, "(objectclass=*)")
+
+	if err != nil {
+		// TODO: we should not give this error out to users
+		http.Error(w, err.Error(), http.StatusInternalServerError)
+		return
+	}
+
+	var users []*msgraph.User
+
+	for _, user := range result.Entries {
+		users = append(
+			users,
+			createUserModelFromLDAP(
+				user,
+			),
+		)
+	}
+
+	render.Status(r, http.StatusOK)
+	render.JSON(w, r, users)
+}
+
+// GetUser implements the Service interface.
+func (g Graph) GetUser(w http.ResponseWriter, r *http.Request) {
+	user := r.Context().Value(userIDKey).(*ldap.Entry)
+
+	render.Status(r, http.StatusOK)
+	render.JSON(w, r, createUserModelFromLDAP(user))
+}
+
+func (g Graph) ldapGetUser(userID string) (*ldap.Entry, error) {
+	conn, err := g.initLdap()
+	if err != nil {
+		return nil, err
+	}
+	filter := fmt.Sprintf("(entryuuid=%s)", userID)
+	result, err := g.ldapSearch(conn, filter)
+	if err != nil {
+		return nil, err
+	}
+	if len(result.Entries) == 0 {
+		return nil, errors.New("user not found")
+	}
+	return result.Entries[0], nil
+}
+
+func (g Graph) initLdap() (*ldap.Conn, error) {
+	g.logger.Info().Msg("Dailing ldap.... ")
+	con, err := ldap.Dial("tcp", "localhost:10389")
+
+	if err != nil {
+		return nil, err
+	}
+
+	if err := con.Bind("cn=admin,dc=example,dc=org", "admin"); err != nil {
+		return nil, err
+	}
+	return con, nil
+}
+
+func (g Graph) ldapSearch(con *ldap.Conn, filter string) (*ldap.SearchResult, error) {
+	search := ldap.NewSearchRequest(
+		"ou=users,dc=example,dc=org",
+		ldap.ScopeWholeSubtree,
+		ldap.NeverDerefAliases,
+		0,
+		0,
+		false,
+		filter,
+		[]string{"dn",
+			"uid",
+			"givenname",
+			"mail",
+			"displayname",
+			"entryuuid",
+			"sn",
+		},
+		nil,
+	)
+
+	return con.Search(search)
+}
+
+func createUserModel(displayName string, id string) *msgraph.User {
```

Comment:
personally I'd move this and `createUserModelFromLDAP` method to an utils file (same package)